e2e/cp_test.go

```diff
@@ -3346,3 +3346,466 @@ func TestCopyRawModeAllowDestinationWithoutPrefix(t *testing.T) {
+		"main.c",
+	}
+
+	for _, filename := range files {
+		putFile(t, s3client, bucket, filename, fileContent)
+	}
+
+	srcpath := fmt.Sprintf("s3://%s", bucket)
+
+	cmd := s5cmd("cp", "--exclude", excludePattern1, "--exclude", excludePattern2, srcpath+"/*", ".")
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: equals("cp %v/main.c main.c", srcpath),
+	})
+
+	// assert s3
+	for _, f := range files {
+		assert.Assert(t, ensureS3Object(s3client, bucket, f, fileContent))
+	}
+
+	expectedFileSystem := []fs.PathOp{
+		fs.WithFile("main.c", fileContent),
+	}
+	// assert local filesystem
+	expected := fs.Expected(t, expectedFileSystem...)
+	assert.Assert(t, fs.Equal(cmd.Dir, expected))
+}
+
+// cp --exclude ".txt" s3://bucket/abc*
+func TestCopyS3ObjectsWithPrefixWithExcludeFilters(t *testing.T) {
+	t.Parallel()
+
+	bucket := s3BucketFromTestName(t)
+
+	s3client, s5cmd, cleanup := setup(t)
+	defer cleanup()
+
+	createBucket(t, s3client, bucket)
+
+	const (
+		excludePattern1 = "*.txt"
+		fileContent     = "content"
+	)
+
+	files := [...]string{
+		"abc/file.txt",
+		"abc/file2.txt",
+		"abc/abc/file3.txt",
+		"abcd/main.py",
+		"ab/file.py",
+		"a/helper.c",
+		"abc.pdf",
+	}
+
+	for _, filename := range files {
+		putFile(t, s3client, bucket, filename, fileContent)
+	}
+
+	srcpath := fmt.Sprintf("s3://%s/abc*", bucket)
+
+	cmd := s5cmd("cp", "--exclude", excludePattern1, srcpath, ".")
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: equals("cp s3://%s/abc.pdf abc.pdf", bucket),
+		1: equals("cp s3://%s/abcd/main.py abcd/main.py", bucket),
+	}, sortInput(true))
+
+	// assert s3
+	for _, f := range files {
+		assert.Assert(t, ensureS3Object(s3client, bucket, f, fileContent))
+	}
+
+	expectedFileSystem := []fs.PathOp{
+		fs.WithFile("abc.pdf", fileContent),
+		fs.WithDir(
+			"abcd",
+			fs.WithFile("main.py", fileContent),
+		),
+	}
+	// assert local filesystem
+	expected := fs.Expected(t, expectedFileSystem...)
+	assert.Assert(t, fs.Equal(cmd.Dir, expected))
+}
+
+// cp --exclude "*.gz" dir/* s3://bucket/
+func TestCopyLocalDirectoryToS3WithExcludeFilter(t *testing.T) {
+	t.Parallel()
+
+	bucket := s3BucketFromTestName(t)
+
+	s3client, s5cmd, cleanup := setup(t)
+	defer cleanup()
+
+	createBucket(t, s3client, bucket)
+
+	folderLayout := []fs.PathOp{
+		fs.WithFile("testfile1.txt", "this is a test file 1"),
+		fs.WithFile("readme.md", "this is a readme file"),
+		fs.WithDir(
+			"a",
+			fs.WithFile("another_test_file.txt", "yet another txt file. yatf."),
+		),
+		fs.WithDir(
+			"b",
+			fs.WithFile("filename-with-hypen.gz", "file has hypen in its name"),
+		),
+	}
+
+	workdir := fs.NewDir(t, "somedir", folderLayout...)
+	defer workdir.Remove()
+
+	const excludePattern = "*.gz"
+
+	src := fmt.Sprintf("%v/", workdir.Path())
+	dst := fmt.Sprintf("s3://%v/prefix/", bucket)
+
+	src = filepath.ToSlash(src)
```

Comment:
`src` where does not end with a wildcard.

Reply:
Maybe we should have three tests for local directories;
* dir (which is why `storage.Stat` call was needed.)
* dir/
* dir/*

Reply:
added these test cases in the other [PR](https://github.com/peak/s5cmd/pull/334).